src/protocol/data/redis/token.c

```diff
@@ -0,0 +1,337 @@
+#include "token.h"
+
+#include "request.h"
+#include "response.h"
+
+#include <cc_define.h>
+#include <cc_print.h>
+#include <cc_util.h>
+
+#include <ctype.h>
+
+#define STR_MAXLEN 255 /* max length for simple string or error */
+#define BULK_MAXLEN (512 * MiB)
+#define ARRAY_MAXLEN (64 * MiB)
+
+#define NIL_STR "$-1\r\n"
+
+
+static inline compose_rstatus_t
+_check_buf_size(struct buf **buf, uint32_t n)
+{
+    while (n > buf_wsize(*buf)) {
+        if (dbuf_double(buf) != CC_OK) {
+            log_debug("failed to write  %u bytes to buf %p: insufficient "
+                    "buffer space", n, *buf);
+
+            return COMPOSE_ENOMEM;
+        }
+    }
+
+    return COMPOSE_OK;
+}
+
+
+static parse_rstatus_t
+_read_str(struct bstring *str, struct buf *buf)
+{
+    /*
+     * Note: buf->rpos is updated in this function, the caller is responsible
+     * for resetting the pointer if necessary.
+     */
+
+    str->len = 0;
+    str->data = buf->rpos;
+    /*
+     * Note: according to @antirez, simple strings are not supposed to be empty.
+     * However, there's no particular harm allowing a null simple string, so
+     * we allow it in this function
+     */
+    for (; buf->rpos < buf->wpos; buf->rpos++) {
+        if (line_end(buf)) {
+            buf->rpos += CRLF_LEN;
+            log_vverb("simple string detected at %p, length %"PRIu32, str->len);
+
+            return PARSE_OK;
+        }
+        if (++str->len > STR_MAXLEN) {
+            log_warn("simple string max length (%d) exceeded", STR_MAXLEN);
+
+            return PARSE_EOVERSIZE;
+        }
+    }
+
+    return PARSE_EUNFIN;
+}
+
+
+static parse_rstatus_t
+_read_int(int64_t *num, struct buf *buf, int64_t min, int64_t max)
+{
+    /*
+     * Note: buf->rpos is updated in this function, the caller is responsible
+     * for resetting the pointer if necessary.
+     */
+    size_t len = 0;
+    int64_t sign = 1;
+
+    if (*buf->rpos == '-') {
+        sign = -1;
+        buf->rpos++;
+    }
+
+    *num = 0;
+    for (; buf_rsize(buf) > 0; buf->rpos++) {
+        if (isdigit(*buf->rpos)) {
+            if (*num < min / 10 || *num > max / 10) {
+                /* TODO(yao): catch the few numbers that will still overflow */
+                log_warn("ill formatted token: integer out of bounds");
+
+                return PARSE_EOVERSIZE;
+            }
+
+            len++;
+            *num = *num * 10ULL + sign * (*buf->rpos - '0');
+        } else {
+            if (len > 0 && line_end(buf)) {
+                buf->rpos += CRLF_LEN;
+                log_vverb("parsed integer, value %"PRIi64, *num);
+
+                return PARSE_OK;
+            } else {
+                if (*buf->rpos == CR) {
```

Comment:
Yea it should. OK I think read_int needs to a bit further tuning, current logic flow is error-prone.